Union Object Initializer ```error: extra brace group at end of initializer```

### What happened?

I have encountered this issue (#10815) or something related in my project with the following example code:

```nim
type
  EmptySeq* {.bycopy.} = object

  ChoiceWithEmptySeq_d* {.bycopy.} = object
    a*: bool

  INNER_C_UNION* {.bycopy, union.} = object
    a*: char
    b*: EmptySeq
    c*: byte
    d*: ChoiceWithEmptySeq_d

  ChoiceWithEmptySeq_selection* = enum
    ChoiceWithEmptySeq_NONE, 
    ChoiceWithEmptySeq_a_PRESENT,
    ChoiceWithEmptySeq_b_PRESENT,  
    ChoiceWithEmptySeq_c_PRESENT,
    ChoiceWithEmptySeq_d_PRESENT

  ChoiceWithEmptySeq* {.bycopy.} = object
    kind*: ChoiceWithEmptySeq_selection
    u*: INNER_C_UNION

  Og_Context* {.bycopy.} = object
    state*: int
    init_done*: bool
    ch*: ChoiceWithEmptySeq
    em*: EmptySeq

var context* : Og_Context = Og_Context(init_done: false)
```

The generated C code attempts to initialize the Object as: `{0, NIM_FALSE, {0, {0, {}, 0, {NIM_FALSE}}}, {}}`
while it should in fact be: `{0, NIM_FALSE, {0, 0}, {}}`. It tries to populate the Union as if it is a regular object, while it should only initialize a single field. 

This bug does not occur when manually initializing the Union:
```nim
var context* : Og_Context = Og_Context(init_done: false, 
                                       ch: ChoiceWithEmptySeq(kind: ChoiceWithEmptySeq_a_PRESENT, 
                                                              u: INNER_C_UNION(a: '\0')))
```


### Nim Version

Nim Compiler Version 1.6.8 [Linux: amd64]
Compiled at 2022-09-27
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: c9f46ca8c9eeca8b5f68591b1abe14b962f80a4c
active boot switches: -d:release

### Current Standard Output Logs

```shell
/home/runner/.cache/nim/main_d/@mmain.nim.c:58:117: error: extra brace group at end of initializer
 text__HGn2aXle70F1hKpo8FMikg = {0, NIM_FALSE, {0, {0, {}, 0, {NIM_FALSE}}}, {}}
                                                       ^
/home/runner/.cache/nim/main_d/@mmain.nim.c:58:117: note: (near initialization for ‘context__HGn2aXle70F1hKpo8FMikg.ch.u’)
/home/runner/.cache/nim/main_d/@mmain.nim.c:58:124: error: extra brace group at end of initializer
 Gn2aXle70F1hKpo8FMikg = {0, NIM_FALSE, {0, {0, {}, 0, {NIM_FALSE}}}, {}}
                                                       ^
/home/runner/.cache/nim/main_d/@mmain.nim.c:58:124: note: (near initialization for ‘context__HGn2aXle70F1hKpo8FMikg.ch.u’)
Error: execution of an external compiler program 'gcc -c  -w   -I/usr/local/lib/nim -I/home/runner/ttv0abnzej -o /home/runner/.cache/nim/main_d/@mmain.nim.c.o /home/runner/.cache/nim/main_d/@mmain.nim.c' failed with exit code: 1
```

### Additional Information

Related Issue: #10815
Platform: Linux Debian 11   
(Kernel Version: 5.10.0-17-amd64)

I've been hitting a similar issue in our nim-eth code base.

Minimized (does still use chronos) code snippit:
```Nim


import chronos

type
  Socket* = object
    a*: int

  ErrorType* = enum
    ErrorA, ErrorB

  Error* = object
    case kind*: ErrorType
    of ErrorA:
      discard
    of ErrorB:
      discard

proc connect(s: Socket): Future[Error] {.async: (raises: [CancelledError]).} =
  discard

let socket = Socket(a: 123)
let res = waitFor socket.connect()
```

Compiles fine with Nim 1.6.20:
```
Nim Compiler Version 1.6.20 [Linux: amd64]
Compiled at 2024-04-07
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: 19fdbfc173bfccb64cb64e0a963e69f52f71fc73
active boot switches: -d:release
```


With Nim devel or 2.0.4

```
/home/deme/.cache/nim/extra_brace_d/@m..@s..@s.nimble@spkgs2@schronos-4.0.2-c736ec5b800fa8df7cf1be9da918578bd5e173d2@schronos@sinternal@sasyncfutures.nim.c:316:266: error: extra brace group at end of initializer
  316 | tuidRCk31w_6 = {{{{{(&NTIv2__1go9brEQ7Per5SINHtx3UOw_)}, {NIM_NIL, NIM_NIL}, {0, NIM_NIL}, {NIM_NIL, NIM_NIL}, NIM_NIL, 0, 0, NIM_NIL, {NIM_NIL, NIM_NIL}}}, {0, {}}}, {}}
      |                                                                                                                                                                  ^

/home/deme/.cache/nim/extra_brace_d/@m..@s..@s.nimble@spkgs2@schronos-4.0.2-c736ec5b800fa8df7cf1be9da918578bd5e173d2@schronos@sinternal@sasyncfutures.nim.c:316:266: note: (near initialization for ‘TM__qMFlZYKlAiPttuidRCk31w_6.Sup.internalValue’)
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3 -pthread   -I'/home/deme/.choosenim/toolchains/nim-#devel/lib' -I/home/deme/repos/nim-eth -o /home/deme/.cache/nim/extra_brace_d/@m..@s..@s.nimble@spkgs2@schronos-4.0.2-c736ec5b800fa8df7cf1be9da918578bd5e173d2@schronos@sinternal@sasyncfutures.nim.c.o /home/deme/.cache/nim/extra_brace_d/@m..@s..@s.nimble@spkgs2@schronos-4.0.2-c736ec5b800fa8df7cf1be9da918578bd5e173d2@schronos@sinternal@sasyncfutures.nim.c' failed with exit code: 1
```

Nim devel:
```
Nim Compiler Version 2.1.1 [Linux: amd64]
Compiled at 2024-06-10
Copyright (c) 2006-2024 by Andreas Rumpf

git hash: 56c95758b2777a36c8608a7b9950ba5ce8ce0b65
active boot switches: -d:release
```

Nim 2.0.4:
```
Nim Compiler Version 2.0.4 [Linux: amd64]
Compiled at 2024-03-28
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: b47747d31844c6bd9af4322efe55e24fefea544c
active boot switches: -d:release
```

nim-chronos revision: [7630f394718ebcdb8577e36faacd78cb7a0b7dd6](https://github.com/status-im/nim-chronos/commit/7630f394718ebcdb8577e36faacd78cb7a0b7dd6)

Note that not using the object variant but instead directly the `ErrorType` makes the compiling succeed.

--- a/compiler/ccgexprs.nim
+++ b/compiler/ccgexprs.nim
@@ -3239,7 +3239,8 @@ proc getNullValueAux(p: BProc; t: PType; obj, constOrNil: PNode,
       getNullValueAux(p, t, it, constOrNil, result, count, isConst, info)
   of nkRecCase:
     getNullValueAux(p, t, obj[0], constOrNil, result, count, isConst, info)
-    if count > 0: result.add ", "
+    var res = ""
+    if count > 0: res.add ", "
     var branch = Zero
     if constOrNil != nil:
       ## find kind value, default is zero if not specified
@@ -3253,18 +3254,21 @@ proc getNullValueAux(p: BProc; t: PType; obj, constOrNil: PNode,
           break
 
     let selectedBranch = caseObjDefaultBranch(obj, branch)
-    result.add "{"
+    res.add "{"
     var countB = 0
     let b = lastSon(obj[selectedBranch])
     # designated initilization is the only way to init non first element of unions
     # branches are allowed to have no members (b.len == 0), in this case they don't need initializer
     if b.kind == nkRecList and not isEmptyCaseObjectBranch(b):
-      result.add "._" & mangleRecFieldName(p.module, obj[0].sym) & "_" & $selectedBranch & " = {"
-      getNullValueAux(p, t,  b, constOrNil, result, countB, isConst, info)
-      result.add "}"
+      res.add "._" & mangleRecFieldName(p.module, obj[0].sym) & "_" & $selectedBranch & " = {"
+      getNullValueAux(p, t,  b, constOrNil, res, countB, isConst, info)
+      res.add "}"
     elif b.kind == nkSym:
-      result.add "." & mangleRecFieldName(p.module, b.sym) & " = "
-      getNullValueAux(p, t,  b, constOrNil, result, countB, isConst, info)
+      res.add "." & mangleRecFieldName(p.module, b.sym) & " = "
+      getNullValueAux(p, t,  b, constOrNil, res, countB, isConst, info)
+    else:
+      return
+    result.add res
     result.add "}"
 
   of nkSym:

